File: thrifty-runtime/src/main/java/com/microsoft/thrifty/protocol/Protocol.java
/*
 * Thrifty
 *
 * Copyright (c) Microsoft Corporation
 *
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING
 * WITHOUT LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE,
 * FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing permissions and limitations under the License.
 */
package com.microsoft.thrifty.protocol;

import com.microsoft.thrifty.transport.Transport;
import okio.ByteString;

import java.io.Closeable;
import java.io.IOException;

public abstract class Protocol implements Closeable {
    protected final Transport transport;

    protected Protocol(Transport transport) {
        if (transport == null) {
            throw new NullPointerException("transport");
        }
        this.transport = transport;
    }

    public abstract void writeMessageBegin(String name, byte typeId, int seqId) throws IOException;

    public abstract void writeMessageEnd() throws IOException;

    public abstract void writeStructBegin(String structName) throws IOException;

    public abstract void writeStructEnd() throws IOException;

    public abstract void writeFieldBegin(String fieldName, int fieldId, byte typeId) throws IOException;

    public abstract void writeFieldEnd() throws IOException;

    public abstract void writeFieldStop() throws IOException;

    public abstract void writeMapBegin(byte keyTypeId, byte valueTypeId, int mapSize) throws IOException;

    public abstract void writeMapEnd() throws IOException;

    public abstract void writeListBegin(byte elementTypeId, int listSize) throws IOException;

    public abstract void writeListEnd() throws IOException;

    public abstract void writeSetBegin(byte elementTypeId, int setSize) throws IOException;

    public abstract void writeSetEnd() throws IOException;

    public abstract void writeBool(boolean b) throws IOException;

    public abstract void writeByte(byte b) throws IOException;

    public abstract void writeI16(short i16) throws IOException;

    public abstract void writeI32(int i32) throws IOException;

    public abstract void writeI64(long i64) throws IOException;

    public abstract void writeDouble(double dub) throws IOException;

    public abstract void writeString(String str) throws IOException;

    public abstract void writeBinary(ByteString buf) throws IOException;

    ////////

    public abstract MessageMetadata readMessageBegin() throws IOException;

    public abstract void readMessageEnd() throws IOException;

    public abstract StructMetadata readStructBegin() throws IOException;

    public abstract void readStructEnd() throws IOException;

    public abstract FieldMetadata readFieldBegin() throws IOException;

    public abstract void readFieldEnd() throws IOException;

    public abstract MapMetadata readMapBegin() throws IOException;

    public abstract void readMapEnd() throws IOException;

    public abstract ListMetadata readListBegin() throws IOException;

    public abstract void readListEnd() throws IOException;

    public abstract SetMetadata readSetBegin() throws IOException;

    public abstract void readSetEnd() throws IOException;

    public abstract boolean readBool() throws IOException;

    public abstract byte readByte() throws IOException;

    public abstract short readI16() throws IOException;

    public abstract int readI32() throws IOException;

    public abstract long readI64() throws IOException;

    public abstract double readDouble() throws IOException;

    public abstract String readString() throws IOException;

    public abstract ByteString readBinary() throws IOException;

    //////////////

    public void flush() throws IOException {
        transport.flush();
    }

    public void reset() {
        // to be implemented by children as needed
    }

    @Override
    public void close() throws IOException {
        this.transport.close();
    }
}

File: thrifty-runtime/src/commonMain/kotlin/com/bendb/thrifty/protocol/Protocol.kt
/*
 * Thrifty
 *
 * Copyright (c) Benjamin Bader
 * Copyright (c) Microsoft Corporation
 *
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING
 * WITHOUT LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE,
 * FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing permissions and limitations under the License.
 */
package com.bendb.thrifty.protocol

import okio.ByteString
import okio.Closeable
import okio.IOException

interface Protocol : Closeable {

    @Throws(IOException::class)
    fun writeMessageBegin(name: String, typeId: Byte, seqId: Int)

    @Throws(IOException::class)
    fun writeMessageEnd()

    @Throws(IOException::class)
    fun writeStructBegin(structName: String)

    @Throws(IOException::class)
    fun writeStructEnd()

    @Throws(IOException::class)
    fun writeFieldBegin(fieldName: String, fieldId: Int, typeId: Byte)

    @Throws(IOException::class)
    fun writeFieldEnd()

    @Throws(IOException::class)
    fun writeFieldStop()

    @Throws(IOException::class)
    fun writeMapBegin(keyTypeId: Byte, valueTypeId: Byte, mapSize: Int)

    @Throws(IOException::class)
    fun writeMapEnd()

    @Throws(IOException::class)
    fun writeListBegin(elementTypeId: Byte, listSize: Int)

    @Throws(IOException::class)
    fun writeListEnd()

    @Throws(IOException::class)
    fun writeSetBegin(elementTypeId: Byte, setSize: Int)

    @Throws(IOException::class)
    fun writeSetEnd()

    @Throws(IOException::class)
    fun writeBool(b: Boolean)

    @Throws(IOException::class)
    fun writeByte(b: Byte)

    @Throws(IOException::class)
    fun writeI16(i16: Short)

    @Throws(IOException::class)
    fun writeI32(i32: Int)

    @Throws(IOException::class)
    fun writeI64(i64: Long)

    @Throws(IOException::class)
    fun writeDouble(dub: Double)

    @Throws(IOException::class)
    fun writeString(str: String)

    @Throws(IOException::class)
    fun writeBinary(buf: ByteString)

    ////////

    @Throws(IOException::class)
    fun readMessageBegin(): MessageMetadata

    @Throws(IOException::class)
    fun readMessageEnd()

    @Throws(IOException::class)
    fun readStructBegin(): StructMetadata

    @Throws(IOException::class)
    fun readStructEnd()

    @Throws(IOException::class)
    fun readFieldBegin(): FieldMetadata

    @Throws(IOException::class)
    fun readFieldEnd()

    @Throws(IOException::class)
    fun readMapBegin(): MapMetadata

    @Throws(IOException::class)
    fun readMapEnd()

    @Throws(IOException::class)
    fun readListBegin(): ListMetadata

    @Throws(IOException::class)
    fun readListEnd()

    @Throws(IOException::class)
    fun readSetBegin(): SetMetadata

    @Throws(IOException::class)
    fun readSetEnd()

    @Throws(IOException::class)
    fun readBool(): Boolean

    @Throws(IOException::class)
    fun readByte(): Byte

    @Throws(IOException::class)
    fun readI16(): Short

    @Throws(IOException::class)
    fun readI32(): Int

    @Throws(IOException::class)
    fun readI64(): Long

    @Throws(IOException::class)
    fun readDouble(): Double

    @Throws(IOException::class)
    fun readString(): String

    @Throws(IOException::class)
    fun readBinary(): ByteString

    //////////////

    @Throws(IOException::class)
    fun flush()

    fun reset() {
        // to be implemented by implementations as needed
    }
}
